src/behavior-considered-undefined.md

```diff
@@ -23,44 +23,56 @@ code.
-  and [`write_unaligned`].
-* Reads of [undef] \(uninitialized) memory.
-* Breaking the [pointer aliasing rules] on accesses through raw pointers;
-  a subset of the rules used by C.
-* `&mut T` and `&T` follow LLVM’s scoped [noalias] model, except if the `&T`
-  contains an [`UnsafeCell<U>`].
-* Mutating non-mutable data &mdash; that is, data reached through a shared
+* Dereferencing (using the `*` operator on) a dangling or unaligned raw pointer.
+* Breaking the [pointer aliasing rules]. `&mut T` and `&T` follow LLVM’s scoped
+  [noalias] model, except if the `&T` contains an [`UnsafeCell<U>`].
+* Mutating non-mutable data (that is, data reached through a shared
   reference or data owned by a `let` binding), unless that data is contained
   within an [`UnsafeCell<U>`].
-* Invoking undefined behavior via compiler intrinsics:
-  * Indexing outside of the bounds of an object with [`offset`] with
-    the exception of one byte past the end of the object.
-  * Using [`std::ptr::copy_nonoverlapping_memory`], a.k.a. the `memcpy32`and
-    `memcpy64` intrinsics, on overlapping buffers.
-* Invalid values in primitive types, even in private fields and locals:
-  * Dangling or null references and boxes.
+* Invoking undefined behavior via compiler intrinsics.
+* Executing code compiled with platform features that the current platform
+  does not support (see [`target_feature`]).
+* Unwinding into another language.
+* Producing an invalid value, even in private fields and locals. "Producing" a
+  value happens any time a value is assigned, passed to a function/primitive
+  operation or returned from a function/primitive operation.
+  The following values are invalid (at their respective type):
   * A value other than `false` (`0`) or `true` (`1`) in a `bool`.
   * A discriminant in an `enum` not included in the type definition.
+  * A null `fn` pointer.
   * A value in a `char` which is a surrogate or above `char::MAX`.
+  * A `!` (all values are invalid for this type).
+  * A dangling or unaligned reference or `Box`, or one that points to an invalid value.
+  * Invalid metadata in a wide reference, `Box` or raw pointer:
+    * slice metadata is invalid if the slice has a total size larger than
+      `isize::MAX` bytes in memory.
+    * `dyn Trait` metadata is invalid if it is not a pointer to a vtable for
+      `Trait` that matches the actual dynamic trait the reference points to.
   * Non-UTF-8 byte sequences in a `str`.
-* Executing code compiled with platform features that the current platform
-  does not support (see [`target_feature`]).
+  * [Uninitialized memory][undef] in the value of an integer (`i*`/`u*`),
+    floating point value (`f*`), or raw pointer.
+  * Invalid values for a type with a custom definition of invalid values, such
+    as a `NonNull` that is null. (Requesting custom invalid values is an
+    unstable feature, but some stable libstd types, like `NonNull`, make use of
```

Comment:
maybe also add `NonZero*` to this list ?

Reply:
Isn't one example enough? I think it is.

Reply:
I thought the only two types in libstd that we expose for which this holds were `NonZero*` and `NonNull`. If there are more then one example is enough, but otherwise we can be complete here.

Reply:
AFAIK those are the only ones -- `NonNull*` is an entire family though, so it's more than two types.

However, I think it is a maintenance hazard to try to list those here.